packages/nimble-components/src/table/specs/table-columns-hld.md

```diff
@@ -0,0 +1,287 @@
+
+### Out of scope of this HLD
+
+Programmatic API for state that could be considered column-centric: width, sort direction, grouped, etc. These concerns should be covered in separate designs covering those topics specifically, allowing for discussion on both the interactive side and the API design on an individual basis.
+
+## Links To Relevant Work Items and Reference Material
+
+[Table Spec](./README.md)
+
+[Table Declarative Columns API prototype:](https://github.com/ni/nimble/blob/325983040e886e52a100664d8fb1129dee767c2f/packages/nimble-components/src/table/tests/table.stories.ts#L23) ([Storybook](https://60e89457a987cf003efc0a5b-twewjutggo.chromatic.com/iframe.html?args=&id=table--table-story&viewMode=story))
+
+## Implementation / Design
+
+Columns will be provided to the table as slotted elements. The slot for the columns will be the default slot for the table, and thus no slot needs to be expressly provided by the user for a column:
+
+```
+<nimble-table>
+    <nimble-table-column-text data-key="name"></nimble-table-column-text>
+    <nimble-table-column-icon data-key="ready"></nimble-table-column-icon>
+    ...
+</nimble-table>
+```
+
+These columns will _not_ have templates/CSS associated with them, and instead be implementations of an interface that will require returning a FAST ViewTemplate for its visual representation. The ordering of the columns in the markup will determine the visual ordering of the columns (top to bottom equals left to right...unless in 'rtl'). Re-ordering of columns will be done, at least at first, through the re-ordering of the columns elements in the DOM.
+
+The table API to support this could look like the following:
+
+```TS
+public Table extends FoundationElement {
+    ...
+
+    @observable
+    public slottedColumns: ITableColumn[] = [];
+
+    private slottedColumnsChanged(): void {
+        if (this.slottedColumns.length > 0) {
+            for (const tableColumn of this.slottedColumns) {
+                if (this.isTableColumn(tableColumn)) {
+                    // do init work
+                }
+            }
+        }
+    }
+}
+
+template:
+<template>
+    <div>
+        ...
+    </div>
+    <slot ${slotted('slottedColumns')}></slot>
+</template>
+```
+
+### `TableCellData` interface
+
+A table cell represents a single column for a single row. The data that a cell has access to will be a subset of the data for the entire row. An instance of a table cell will be generic to describe the subset of data it contains, where `TCellData` is a subset of the data represented by `TableRowData`:
+
+```TS
+interface TableCellData<TCellData extends TableRowData> {
```

Comment:
Doesn't this mean that `TCellData` must be a subclass of `TableRowData`? Seems like that wouldn't make it a subset of the row data. What am I missing?

Reply:
Extending generic type parameters just means that `TCellData` must at least match the type definition of `TableRowData`. Ultimately, it will be beneficial to provide links to the `TableRowData` definition. 

So, while the type of `TCellData` is a technical superset of the `TableRowData` type, in practice, the collection of values that `TCellData` represents is a subset of the values that a `TableRowData` instance would be aware of (which would effectively be a collection of `TCellData` values).